File: packages/pvm-assemblyscript/assembly/host-functions/general/poke.ts
import { RESULT_CODE_PANIC } from '../../config'
import {
  ACCUMULATE_ERROR_OK,
  ACCUMULATE_ERROR_OOB,
  ACCUMULATE_ERROR_WHO,
  HostFunctionResult,
} from '../accumulate/base'
import {
  HostFunctionContext,
  HostFunctionParams,
  PeekPokeParams,
  PVMGuest,
  RefineInvocationContext,
} from './base'
import { BaseHostFunction } from './base'

/**
 * POKE host function (Ω_O)
 *
 * Writes memory to a PVM machine instance
 *
 * *** GRAY PAPER FORMULA ***
 * Gray Paper: pvm_invocations.tex, Ω_O (poke = 10)
 *
 * Parameters: [n, s, o, z] = registers[7:4]
 * - n: machine ID
 * - s: source offset in current memory
 * - o: destination offset in machine's memory
 * - z: length
 *
 * Return states (equation 586-593):
 * - panic when Nrange{s}{z} not ⊆ readable[memory]  (source not readable)
 * - continue with WHO when n not ∈ keys(m)  (machine doesn't exist)
 * - continue with OOB when Nrange{o}{z} not ⊆ writable{m[n].ram}  (destination not writable)
 * - continue with OK otherwise
 *
 * Memory update: (m'[n].ram)[o:z] = mem[s:z]
 */
export class PokeHostFunction extends BaseHostFunction {
  functionId: u64 = u64(10) // POKE function ID
  name: string = 'poke'
  gasCost: u64 = u64(10)

  execute(
    context: HostFunctionContext,
    params: HostFunctionParams | null,
  ): HostFunctionResult {
    if (!params) {
      return new HostFunctionResult(255) // continue execution
    }
    const pokeParams = params as PeekPokeParams
    if (!pokeParams.refineContext) {
      context.registers[7] = ACCUMULATE_ERROR_WHO
      return new HostFunctionResult(RESULT_CODE_PANIC)
    }

    // Gray Paper: [n, s, o, z] = registers[7:4]
    const machineId = u64(context.registers[7])
    const sourceOffset = u64(context.registers[8]) // s: source
    const destOffset = u64(context.registers[9]) // o: destination
    const length = u64(context.registers[10]) // z: length

    // Gray Paper error check order:
    // 1. Check if source range is readable in current memory → panic
    // Read data from current PVM's memory
    const readResult_data = context.ram.readOctets(
      u32(sourceOffset),
      u32(length),
    )
    const data = readResult_data.data
    const readFaultAddress = readResult_data.faultAddress
    if (data === null || readFaultAddress !== 0) {
      return new HostFunctionResult(RESULT_CODE_PANIC)
    }

    // 2. Check if machine exists → WHO
    const refineContext = pokeParams.refineContext!
    const machine = this.getPVMMachine(refineContext, machineId)
    if (!machine) {
      context.registers[7] = ACCUMULATE_ERROR_WHO
      return new HostFunctionResult(255) // continue
    }

    // 3. Check if destination range is writable → OOB
    // Gray Paper: (m'[n].ram)[o:z] = mem[s:z]
    const writeResult = machine.pvm.state.ram.writeOctets(u32(destOffset), data)
    if (writeResult.hasFault) {
      context.registers[7] = ACCUMULATE_ERROR_OOB
      return new HostFunctionResult(255) // continue (not HALT)
    }

    // Return OK (0) for success
    context.registers[7] = ACCUMULATE_ERROR_OK

    return new HostFunctionResult(255) // continue execution
  }

  getPVMMachine(
    refineContext: RefineInvocationContext,
    machineId: u64,
  ): PVMGuest | null {
    if (refineContext.machines.has(machineId)) {
      return refineContext.machines.get(machineId)!
    }
    return null
  }
}

File: packages/pvm-rust/src/host_functions/general/poke.rs
//! POKE host function (Ω_O). Gray Paper: function ID 10.
//! r7=machine ID, r8=source offset, r9=dest offset in machine, r10=length.
//! No refineContext → set r7=WHO, return PANIC.

use crate::config::{FUNC_POKE, REG_OK, REG_OOB, REG_WHO};
use crate::host_functions::base::{HostFunction, HostFunctionContext, HostFunctionResult};

/// POKE (10): with refine_context: read from current RAM, write to machine RAM; source unreadable → PANIC; WHO if no machine, OOB if dest unwritable.
pub struct PokeHostFunction;

impl HostFunction for PokeHostFunction {
    fn function_id(&self) -> u8 {
        FUNC_POKE
    }
    fn name(&self) -> &'static str {
        "poke"
    }
    fn execute(&self, context: &mut HostFunctionContext<'_>) -> HostFunctionResult {
        let Some(refine) = context.refine_context.as_mut() else {
            context.registers[7] = REG_WHO;
            crate::host_log!("[hostfn] poke PANIC: no refine_context");
            return HostFunctionResult::panic();
        };

        let machine_id = context.registers[7];
        let source_offset = context.registers[8] as u32;
        let dest_offset = context.registers[9] as u32;
        let length = context.registers[10] as u32;

        let read_result = context.ram.read_octets(source_offset, length);
        let Some(data) = read_result.data else {
            crate::host_log!(
                "[hostfn] poke PANIC: source read returned no data (offset={}, len={})",
                source_offset, length
            );
            return HostFunctionResult::panic();
        };
        if read_result.fault_address != 0 {
            crate::host_log!(
                "[hostfn] poke PANIC: source read fault (offset={}, len={}, fault_address={})",
                source_offset, length, read_result.fault_address
            );
            return HostFunctionResult::panic();
        }

        let mut write_ok = None;
        let found = refine.with_machine(machine_id, &mut |machine| {
            write_ok = Some(
                machine.ram_is_writable(dest_offset, length) && machine.ram_write(dest_offset, &data),
            );
        });

        if !found {
            context.registers[7] = REG_WHO;
            HostFunctionResult::continue_execution()
        } else if write_ok == Some(true) {
            context.registers[7] = REG_OK;
            HostFunctionResult::continue_execution()
        } else {
            context.registers[7] = REG_OOB;
            HostFunctionResult::continue_execution()
        }
    }
}
